packages/xstate-react/src/useActor.ts

```diff
@@ -0,0 +1,69 @@
+import { useState, useEffect, useMemo, useRef, useCallback } from 'react';
+import { ActorRef, Sender, ActorRefLike } from './types';
+import { EventObject, Interpreter, Actor } from 'xstate';
+import { fromService } from './useService';
+
+function resolveActor(
+  actorLike:
+    | ActorRef<any, any>
+    | ActorRefLike<any, any>
+    | Interpreter<any, any>
+    | Actor
+): ActorRef<any, any> & { deferred?: boolean } {
+  if (actorLike instanceof Interpreter) {
+    return fromService(actorLike);
+  }
+
+  if (!('current' in actorLike)) {
+    return {
+      stop: () => {
+        /* do nothing */
+      },
+      ...actorLike,
+      subscribe: actorLike.subscribe as any, // TODO: fix (subscribe will work, just improperly typed)
+      name: actorLike.id!,
+      current: actorLike.state!
+    };
+  }
+
+  return actorLike;
+}
+
+export function useActor<TEvent extends EventObject, TEmitted = any>(
+  actorLike: ActorRefLike<TEvent, TEmitted> | Actor
+): [TEmitted, Sender<TEvent>] {
+  const actor = useMemo(() => resolveActor(actorLike), [actorLike]);
+  const deferredEventsRef = useRef<EventObject[]>([]);
```

Comment:
right now this is owned by a hook, but not by an actor (or in other words - there is no direct mapping between this array and an actor), so there is a potential of an actor changing in between flushing or sending to it and some events might end up being sent to a new actor, whereas all scheduled events should only target the one that was here at the time when `send` got called

Reply:
> so there is a potential of an actor changing in between flushing or sending to it and some events might end up being sent to a new actor

That is exactly the point of `deferredEventsRef` - we _want_ to send deferred events to the _new_ live actor.

Reply:
Right, I was commenting about a different case - but it's hard to solve with the current design and swapping actors can be also considered an anti-pattern, so nevermind for now.

Reply:
We'll re-examine this for V5.